File: python/test/exchange/test_trade.py
# -*- coding: utf-8 -*-

__all__ = ['test_trade']


def test_trade(exchange, trade, method, symbol):

    trade_keys = [
        'info',
        'id',
        'timestamp',
        'datetime',
        'symbol',
        'order',
        'type',
        'side',
        'takerOrMaker',
        'price',
        'amount',
        'cost',
    ]

    for key in trade_keys:
        assert key in trade

    assert ('fee' in trade) or ('fees' in trade)

    fee_keys = ['cost', 'currency']

    if ('fee' in trade) and (trade['fee'] is not None):
        assert isinstance(trade['fee'], dict)
        for key in fee_keys:
            assert key in trade['fee']

    if ('fees' in trade) and (trade['fees'] is not None):
        assert isinstance(trade['fees'], list)
        for fee in trade['fees']:
            for key in fee_keys:
                assert key in fee

    assert (trade['id'] is None) or isinstance(trade['id'], str)
    assert (trade['timestamp'] is None) or isinstance(trade['timestamp'], int)

    if trade['timestamp'] is not None:
        assert trade['timestamp'] > 1230940800000  # 03 Jan 2009 - first block
        assert trade['timestamp'] < 2147483648000  # 19 Jan 2038 - int32 overflows

    # //------------------------------------------------------------------

    # const adjustedNow = now + 60000

    # const exchangesExcludedFromTimestampCheck = [
    #     'coinbasepro',
    #     'coinbaseprime', // ... as well, probably
    #     'kucoin2',
    # ]

    # if (trade.timestamp) {
    #     if (!exchangesExcludedFromTimestampCheck.includes (exchange.id)) {
    #         assert (trade.timestamp < adjustedNow, 'trade.timestamp is greater than or equal to current time: trade: ' + exchange.iso8601 (trade.timestamp) + ' now: ' + exchange.iso8601 (now))
    #     }
    # }

    # //------------------------------------------------------------------

    assert trade['datetime'] == exchange.iso8601(trade['timestamp'])

    # const isExchangeLackingFilteringTradesBySymbol = [
    #     'kraken', // override for kraken and possibly other exchanges as well, can't return private trades per symbol at all
    # ].includes (exchange.id)
    # if (!isExchangeLackingFilteringTradesBySymbol)
    #     assert (trade.symbol === symbol, 'trade symbol is not equal to requested symbol: trade: ' + trade.symbol + ' reqeusted: ' + symbol)

    assert (trade['type'] is None) or isinstance(trade['type'], str)
    assert (trade['side'] is None) or (trade['side'] == 'buy') or (trade['side'] == 'sell')
    assert (trade['order'] is None) or isinstance(trade['order'], str)
    assert isinstance(trade['price'], float)
    assert trade['price'] > 0
    assert isinstance(trade['amount'], float)
    assert trade['amount'] >= 0
    assert (trade['takerOrMaker'] is None) or (trade['takerOrMaker'] == 'taker') or (trade['takerOrMaker'] == 'maker')

File: ts/src/test/Exchange/test.trade.ts

import testSharedMethods from './test.sharedMethods.js';

function testTrade (exchange, method, entry, symbol, now) {
    const format = {
        'info': { },
        'id': '12345-67890:09876/54321', // string trade id
        'timestamp': 1502962946216, // Unix timestamp in milliseconds
        'datetime': '2017-08-17 12:42:48.000', // ISO8601 datetime with milliseconds
        'symbol': 'ETH/BTC', // symbol
        'order': '12345-67890:09876/54321', // string order id or undefined/None/null
        'side': 'buy', // direction of the trade, 'buy' or 'sell'
        'takerOrMaker': 'taker', // string, 'taker' or 'maker'
        'price': exchange.parseNumber ('0.06917684'), // float price in quote currency
        'amount': exchange.parseNumber ('1.5'), // amount of base currency
        'cost': exchange.parseNumber ('0.10376526'), // total cost (including fees), `price * amount`
    };
    // todo: add takeOrMaker as mandatory set
    const emptyNotAllowedFor = [ 'side', 'price', 'amount', 'cost' ];
    testSharedMethods.assertStructureKeys (exchange, method, entry, format, emptyNotAllowedFor);
    testSharedMethods.assertCommonTimestamp (exchange, method, entry, now);
    testSharedMethods.assertSymbol (exchange, method, entry, 'symbol', symbol);
    //
    testSharedMethods.assertAgainstArray (exchange, method, entry, 'side', [ 'buy', 'sell' ]);
    testSharedMethods.assertAgainstArray (exchange, method, entry, 'takerOrMaker', [ 'taker', 'maker' ]);
    testSharedMethods.reviseFeeObject (exchange, method, entry['fee']);
    testSharedMethods.reviseFeesObject (exchange, method, entry['fees']);
}

export default testTrade;
